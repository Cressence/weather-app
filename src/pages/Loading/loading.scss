@import './../../assets/scss/variables.scss';

.loading-container {
    display: flex;
    justify-content: center;
    top: 30%;
    height: 100%;
    position: relative;

    .loader-text {
        text-align: center;
        font-size: $sub-title;
        font-weight: bolder;
        letter-spacing: 3px;
    }
    .back-button {
      width: 150px;
      margin: auto 40%;
    }

    .loader {
        position: absolute;
        margin-left: 10px;

        .dot {
            width: 20px;
            height: 20px;
            background: $dark-shade1;
            display: inline-block;
            border-radius: 50%;
            margin-left: 2px;
            margin-right: 2px;
            animation: vertical-wave 2s infinite;

            @for $i from 1 through length($animation-colors) {
                &:nth-child(#{length($animation-colors)}n + #{$i}) {
                  background: nth($animation-colors, $i);
                  animation-delay: 0.2s * $i;
                }
              }
        }

        @media (max-width: $medium-devices){
            margin-left: 15px;
        }
    }

    @keyframes vertical-wave {
        0% {
          transform: translate(0, 0px);
        }
        50% {
          opacity: 0.3;
          transform: translate(0, 20px);
        }
        100% {
          transform: translate(0, 0px);
        }
      }
}