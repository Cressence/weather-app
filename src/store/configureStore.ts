import { Store, createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from './root/reducers/rootReducer';

export const store = configureStore();

function configureStore(): Store {
    const composeEnhancers = (window && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

    // Build middleware
    const middleware = composeEnhancers(applyMiddleware(thunk));

    //Create store
    const store = createStore(rootReducer, {}, middleware);

    return store;
}