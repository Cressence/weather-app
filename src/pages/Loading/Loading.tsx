import { useEffect } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { Dispatch } from "redux";
import { useHistory } from "react-router-dom";
import { Button } from '@material-ui/core';

import Wrapper from './../../components/Wrapper';
import { RootState } from '../../store/root/rootReducer';
import { getWeatherData, resetData } from "./../../store/root/action";
import './loading.scss';

const Loading = ( ) => {
    const selectedUnit = localStorage.getItem('unit');
    const selectedCity = localStorage.getItem('city');
    const { weatherInfo, appError } = useSelector((state: RootState) => ({
        weatherInfo: state.weatherInfo.weatherData,
        appError: state.weatherInfo.appError
    }));

    const dispatch: Dispatch<any> = useDispatch();
    const history = useHistory();

    const goBackToDefaultWeather = () => {
        localStorage.clear();
        dispatch(resetData());
    }

    useEffect(() => {
        const MAIN_ROUTE = process.env.REACT_APP_MAIN_ROUTE;
        dispatch(
            getWeatherData(
                selectedCity === null? 'Munich': selectedCity,
                selectedUnit === null ? 'fahrenheit': selectedUnit
            )
        );

        if (weatherInfo !== null) {
            history.push(`${MAIN_ROUTE}/info`);
        }
    }, [dispatch, history, weatherInfo, selectedUnit, selectedCity]);

    return (
        <Wrapper>
            <div className='loading-container'>
                {
                    appError !== null ?
                    <div>
                        <p className='loader-text'>City "{selectedCity}" not found!</p>
                        <Button onClick={goBackToDefaultWeather} className='back-button' variant="contained" color="primary">Back</Button>
                    </div>             
                    :
                    <div>
                    <p className='loader-text'>Almost there</p>
                    <div className="loader">
                        <div className="dot" />
                        <div className="dot" />
                        <div className="dot" />
                        <div className="dot" />
                        <div className="dot" />
                    </div>
                </div>
                }
                
                
            </div>
        </Wrapper>
    );
}

export default Loading;