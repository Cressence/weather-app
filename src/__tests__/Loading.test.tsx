// import { cleanup, render, screen } from '@testing-library/react';
import { shallow, mount } from 'enzyme';
import { ReactElement } from "react";
import {rest} from 'msw';
import {setupServer} from 'msw/node';
import { Provider } from "react-redux";
import {store} from "./../store/configureStore";
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

import  Loading from '../pages/Loading/Loading';

describe('<Loading />: Loading Screen', () => {
    
    it('shows loader while data is loading', () => {
            const component = mount(
                <Provider store={store} >
                    <Loading />
                </Provider>
            );
            
        const loaderText = component.find('p.loader-text').getDOMNode();
                
        expect(component).toMatchSnapshot();
        expect(loaderText.innerHTML).toBe('Almost there');
    });

    it('displays error if unknown city is sent', (done) => {
        let component =mount(
            <Provider store={store} >
                    <Loading />
            </Provider>,
        );

        // const loaderText = component.find('p.loader-text').getDOMNode();
        // expect(loaderText.innerHTML).toBe('Almost there');

        // const mockResponse = (status:number, response:any) => {
        //     return new Response(response, {
        //         status: status,
        //         headers: {
        //             'Content-type': 'application/json'
        //         }
        //     });
        // }
        
        // setTimeout(() => {
        //     component.update();
        //     console.log(component.debug());
        //     done();
        //   }, 10000)
        // async (done:any) => {
        //     const fakePromise = Promise.resolve(mockResponse(
        //         404,
        //         JSON.stringify({message: 'city not found'})
        //      ));
        //      window.fetch = jest.fn().mockImplementationOnce(() => {
        //          return fakePromise
        //      });
        //      console.log('hey');
             
        //     //  expect.assertions(2);
        //      const wrapper = mount(<Loading  />);
        //      await Promise.all([fakePromise]);
        //      setImmediate(() => {
        //          try {
        //              console.log(wrapper.find('p.loader-text').getDOMNode().innerHTML);
                     
        //             expect(wrapper.find('p.loader-text').getDOMNode().innerHTML).toContain('not jnkhj!')
        //          } catch (e) {
        //              done.fail(e);
        //          }
        //          done();
        //      });
        //     }
    });
});