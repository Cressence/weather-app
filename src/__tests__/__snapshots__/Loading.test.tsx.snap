// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading />: Loading Screen shows loader while data is loading 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Loading>
    <Wrapper>
      <div
        className="wrapper-container"
      >
        <WithStyles(ForwardRef(Card))
          className="wrapper-content"
        >
          <ForwardRef(Card)
            className="wrapper-content"
            classes={
              Object {
                "root": "MuiCard-root",
              }
            }
          >
            <WithStyles(ForwardRef(Paper))
              className="MuiCard-root wrapper-content"
              elevation={1}
            >
              <ForwardRef(Paper)
                className="MuiCard-root wrapper-content"
                classes={
                  Object {
                    "elevation0": "MuiPaper-elevation0",
                    "elevation1": "MuiPaper-elevation1",
                    "elevation10": "MuiPaper-elevation10",
                    "elevation11": "MuiPaper-elevation11",
                    "elevation12": "MuiPaper-elevation12",
                    "elevation13": "MuiPaper-elevation13",
                    "elevation14": "MuiPaper-elevation14",
                    "elevation15": "MuiPaper-elevation15",
                    "elevation16": "MuiPaper-elevation16",
                    "elevation17": "MuiPaper-elevation17",
                    "elevation18": "MuiPaper-elevation18",
                    "elevation19": "MuiPaper-elevation19",
                    "elevation2": "MuiPaper-elevation2",
                    "elevation20": "MuiPaper-elevation20",
                    "elevation21": "MuiPaper-elevation21",
                    "elevation22": "MuiPaper-elevation22",
                    "elevation23": "MuiPaper-elevation23",
                    "elevation24": "MuiPaper-elevation24",
                    "elevation3": "MuiPaper-elevation3",
                    "elevation4": "MuiPaper-elevation4",
                    "elevation5": "MuiPaper-elevation5",
                    "elevation6": "MuiPaper-elevation6",
                    "elevation7": "MuiPaper-elevation7",
                    "elevation8": "MuiPaper-elevation8",
                    "elevation9": "MuiPaper-elevation9",
                    "outlined": "MuiPaper-outlined",
                    "root": "MuiPaper-root",
                    "rounded": "MuiPaper-rounded",
                  }
                }
                elevation={1}
              >
                <div
                  className="MuiPaper-root MuiCard-root wrapper-content MuiPaper-elevation1 MuiPaper-rounded"
                >
                  <WithStyles(ForwardRef(CardContent))>
                    <ForwardRef(CardContent)
                      classes={
                        Object {
                          "root": "MuiCardContent-root",
                        }
                      }
                    >
                      <div
                        className="MuiCardContent-root"
                      >
                        <div
                          className="loading-container"
                        >
                          <div>
                            <p
                              className="loader-text"
                            >
                              Almost there
                            </p>
                            <div
                              className="loader"
                            >
                              <div
                                className="dot"
                              />
                              <div
                                className="dot"
                              />
                              <div
                                className="dot"
                              />
                              <div
                                className="dot"
                              />
                              <div
                                className="dot"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </ForwardRef(CardContent)>
                  </WithStyles(ForwardRef(CardContent))>
                </div>
              </ForwardRef(Paper)>
            </WithStyles(ForwardRef(Paper))>
          </ForwardRef(Card)>
        </WithStyles(ForwardRef(Card))>
      </div>
    </Wrapper>
  </Loading>
</Provider>
`;
